contrib/golang/filters/http/source/common/dso/dso.cc

```diff
@@ -0,0 +1,159 @@
+#include "contrib/golang/filters/http/source/common/dso/dso.h"
+
+#include "source/common/common/assert.h"
+
+namespace Envoy {
+namespace Dso {
+
+bool DsoInstanceManager::load(std::string dso_id, std::string dso_name) {
+  ENVOY_LOG_MISC(debug, "load {} {} dso instance.", dso_id, dso_name);
+  if (getDsoInstanceByID(dso_id) != nullptr) {
+    return true;
+  }
+
+  DsoStoreType& dsoStore = DsoInstanceManager::getDsoStore();
+  absl::WriterMutexLock lock(&dsoStore.mutex_);
+  DsoInstancePtr dso(new DsoInstance(dso_name));
+  if (!dso->loaded()) {
+    return false;
+  }
+  dsoStore.map_[dso_id] = std::move(dso);
+  return true;
+}
+
+bool DsoInstanceManager::unload(std::string dso_id) {
+  ENVOY_LOG_MISC(debug, "unload {} dso instance.", dso_id);
+  DsoStoreType& dsoStore = DsoInstanceManager::getDsoStore();
+  absl::WriterMutexLock lock(&dsoStore.mutex_);
+  return dsoStore.map_.erase(dso_id) == 1;
+}
+
+DsoInstancePtr DsoInstanceManager::getDsoInstanceByID(std::string dso_id) {
+  DsoStoreType& dsoStore = DsoInstanceManager::getDsoStore();
+  absl::ReaderMutexLock lock(&dsoStore.mutex_);
+  auto it = dsoStore.map_.find(dso_id);
+  if (it != dsoStore.map_.end()) {
+    return it->second;
+  }
+
+  return nullptr;
+}
+
+std::string DsoInstanceManager::show() {
+  DsoStoreType& dsoStore = DsoInstanceManager::getDsoStore();
+  absl::ReaderMutexLock lock(&dsoStore.mutex_);
+  std::string ids = "";
+  for (auto& it : dsoStore.map_) {
+    ids += it.first;
+    ids += ",";
+  }
+  return ids;
+}
+
+DsoInstance::DsoInstance(const std::string dso_name) : dso_name_(dso_name) {
+  ENVOY_LOG_MISC(info, "loading symbols from so file: {}", dso_name);
+
+  handler_ = dlopen(dso_name.c_str(), RTLD_LAZY);
+  if (!handler_) {
+    ENVOY_LOG_MISC(error, "cannot load : {} error: {}", dso_name, dlerror());
+    return;
+  }
+
+  loaded_ = true;
+
+  auto func = dlsym(handler_, "envoyGoFilterNewHttpPluginConfig");
+  if (func) {
+    envoy_go_filter_new_http_plugin_config_ =
+        reinterpret_cast<GoUint64 (*)(GoUint64 p0, GoUint64 p1)>(func);
+  } else {
+    loaded_ = false;
+    ENVOY_LOG_MISC(error, "lib: {}, cannot find symbol: envoyGoFilterNewHttpPluginConfig, err: {}",
+                   dso_name, dlerror());
+  }
```

Comment:
nit: I would add a helper for this repetition.

Reply:
Done. Thanks for your comments.